### PR TITLE
fix: common_subexpr_eliminate rule should not apply to short-circuit expression

### LINKED ISSUE #8874: Shouldn't extract common sub expression for  short-circuited operation.
### Describe the bug

As @jackwener said in https://github.com/apache/arrow-datafusion/pull/8833#discussion_r1452107106, we shouldn't extract common sub-expressions for short-circuited operations, such as COALESCE | OR.
In #8833, we deal with `CASE WHEN,` but all short-circuited expressions shouldn't be extracted from common sub-expressions.

### To Reproduce

_No response_

### Expected behavior

Shouldn't extract common sub-expressions for short-circuited operation.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by haohuaijin**
## Which issue does this PR close?

Closes #8874

## Rationale for this change

see #8874

## What changes are included in this PR?
before this pr
```
❯ explain select coalesce(1, y/x), coalesce(2, y/x) from t;
+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                  |
+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: coalesce(Int64(1), CAST(t.y / t.x AS Int64)t.y / t.xt.xt.y AS t.y / t.x), coalesce(Int64(2), CAST(t.y / t.x AS Int64)t.y / t.xt.xt.y AS t.y / t.x)                                        |
|               |   Projection: CAST(t.y / t.x AS Int64) AS CAST(t.y / t.x AS Int64)t.y / t.xt.xt.y                                                                                                                     |
|               |     TableScan: t projection=[x, y]                                                                                                                                                                    |
| physical_plan | ProjectionExec: expr=[coalesce(1, CAST(t.y / t.x AS Int64)t.y / t.xt.xt.y@0) as coalesce(Int64(1),t.y / t.x), coalesce(2, CAST(t.y / t.x AS Int64)t.y / t.xt.xt.y@0) as coalesce(Int64(2),t.y / t.x)] |
|               |   ProjectionExec: expr=[CAST(y@1 / x@0 AS Int64) as CAST(t.y / t.x AS Int64)t.y / t.xt.xt.y]                                                                                                          |
|               |     MemoryExec: partitions=1, partition_sizes=[1]                                                                                                                                                     |
|               |                                                                                                                                                                                                       |
+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set. Query took 0.014 seconds.
```
after this pr
```
❯ explain select coalesce(1, y/x), coalesce(2, y/x) from t;
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: coalesce(Int64(1), CAST(t.y / t.x AS Int64)), coalesce(Int64(2), CAST(t.y / t.x AS Int64))                                                              |
|               |   TableScan: t projection=[x, y]                                                                                                                                    |
| physical_plan | ProjectionExec: expr=[coalesce(1, CAST(y@1 / x@0 AS Int64)) as coalesce(Int64(1),t.y / t.x), coalesce(2, CAST(y@1 / x@0 AS Int64)) as coalesce(Int64(2),t.y / t.x)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                                                                                     |
|               |                                                                                                                                                                     |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set. Query took 0.008 seconds.
```

But due to the reason describe in #8927, the above two query will get same result(get drive by zero error).

## Are these changes tested?

yes, add some negative test

### CODE REVIEW

File: datafusion/sqllogictest/test_files/select.slt [hunk 1]
```diff
@@ -1129,5 +1129,35 @@ FROM t AS A, (SELECT * FROM t WHERE x = 0) AS B;
 0 0
 0 0
 
+query TT
+explain select coalesce(1, y/x), coalesce(2, y/x) from t;
```
Comment (haohuaijin) [inline datafusion/sqllogictest/test_files/select.slt:1133]: because of the reason describe in #8927, use plan to test.

**Comment by haohuaijin** [comment]
cc @jackwener @liukun4515 @alamb

File: datafusion/optimizer/src/common_subexpr_eliminate.rs [hunk 2]
```diff
@@ -655,6 +658,20 @@ impl TreeNodeVisitor for ExprIdentifierVisitor<'_> {
     }
 }
 
+/// Check if the expression is short-circuit expression
+fn is_short_circuit_expression(expr: &Expr) -> bool {
```
Comment (alamb) [inline datafusion/optimizer/src/common_subexpr_eliminate.rs:662]: I think we should put this method on expr so it is easier to find and potentially use 

```rust
impl Expr {
...
  /// returns true if some of this `exprs` subexpressions may not be evaluated
  /// and thus any side effects (like divide by zero) may not be encountered
  fn short_circuits(&self) -> bool { 
   ...
   }
```

File: datafusion/sqllogictest/test_files/select.slt (hunk 1 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/select.slt:1133]: Can we please add some comments here explaining the rationale and what the expected outputs are (so future readers know if changes are expected)

Something like  this perhaps:

```
# Expressions that short circuit should not be refactored out as that may cause side effects (divide by zero)
# at plan time that would not actually happen during execution
```


Also, can you please add the tests that now pass (e.g. `select coalesce(1, y/x), coalesce(2, y/x) from t;`) so if someone breaks this code by accident, those queries would start failing, which might be easier to quickly tell is incorrect

**Review by alamb — COMMENTED** [review]
Thank you for this PR @haohuaijin  -- the code looks good to me 🙏 


I also took a look at https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.BuiltinScalarFunction.html# and https://docs.rs/datafusion/latest/datafusion/logical_expr/expr/enum.Expr.html and didn't see any additional functions that shoudl be marked as short circuiting. 

I had a code structure suggestion that I don't think is needed (though I think it would improve this PR). 

I think a few more tests that actually run (and don't error) should be added prior to merge.

Thanks again!

**Review by jackwener — APPROVED** [review]
Thanks @haohuaijin @alamb .

@alamb 's review suggestions are already very detailed, and I agree with his opinion. After those impromentent, we can merge this PR

File: datafusion/optimizer/src/common_subexpr_eliminate.rs (hunk 2 above)
Comment (jackwener) [reply datafusion/optimizer/src/common_subexpr_eliminate.rs:662]: Additionally, we'd better enumerate all expressions here.
This means we would need to consider whether it's short-circuiting when adding a new expression.

File: datafusion/sqllogictest/test_files/select.slt [hunk 3]
```diff
@@ -1129,5 +1129,49 @@ FROM t AS A, (SELECT * FROM t WHERE x = 0) AS B;
...
+EXPLAIN SELECT y = 0 or 1 / y < 1, x = 0 or y = 0 or 1 / y < 1 / x from t;
+----
+logical_plan
+Projection: t.y = Int32(0) OR Int64(1) / CAST(t.y AS Int64) < Int64(1) AS t.y = Int64(0) OR Int64(1) / t.y < Int64(1), t.x = Int32(0) OR t.y = Int32(0) OR Int64(1) / CAST(t.y AS Int64) < Int64(1) / CAST(t.x AS Int64) AS t.x = Int64(0) OR t.y = Int64(0) OR Int64(1) / t.y < Int64(1) / t.x
+--TableScan: t projection=[x, y]
+physical_plan
+ProjectionExec: expr=[y@1 = 0 OR 1 / CAST(y@1 AS Int64) < 1 as t.y = Int64(0) OR Int64(1) / t.y < Int64(1), x@0 = 0 OR y@1 = 0 OR 1 / CAST(y@1 AS Int64) < 1 / CAST(x@0 AS Int64) as t.x = Int64(0) OR t.y = Int64(0) OR Int64(1) / t.y < Int64(1) / t.x]
+--MemoryExec: partitions=1, partition_sizes=[1]
+
+# due to the reason describe in https://github.com/apache/arrow-datafusion/issues/8927,
```
Comment (haohuaijin) [inline datafusion/sqllogictest/test_files/select.slt:1165]: I'm unsure if those tests are appropriate.

File: datafusion/sqllogictest/test_files/select.slt [hunk 4]
```diff
@@ -1129,5 +1129,49 @@ FROM t AS A, (SELECT * FROM t WHERE x = 0) AS B;
...
+# the following queries will fail
+query error
+select coalesce(1, y/x), coalesce(2, y/x) from t;
+
+query error
+SELECT y > 0 and 1 / y < 1, x > 0 and y > 0 and 1 / y < 1 / x from t;
```
Comment (jackwener) [inline datafusion/sqllogictest/test_files/select.slt:1171]: I test this SQL by using this PR branch:

I find it's different

```sql
❯ SELECT y > 0 and 1 / y < 1, x > 0 and y > 0 and 1 / y < 1 / x from t;

\+----------------------------------------------+-----------------------------------------------------------------------+
| t.y > Int64(0) AND Int64(1) / t.y < Int64(1) | t.x > Int64(0) AND t.y > Int64(0) AND Int64(1) / t.y < Int64(1) / t.x |
+----------------------------------------------+-----------------------------------------------------------------------+
| true                                         | true                                                                  |
| true                                         | true                                                                  |
+----------------------------------------------+-----------------------------------------------------------------------+
2 rows in set. Query took 0.006 seconds.
```

File: datafusion/sqllogictest/test_files/select.slt (hunk 3 above)
Comment (jackwener) [reply datafusion/sqllogictest/test_files/select.slt:1165]: It's related with https://github.com/apache/arrow-datafusion/issues/8910

File: datafusion/expr/src/expr.rs [hunk 5]
```diff
@@ -1266,6 +1266,54 @@ impl Expr {
...
+    pub fn short_circuits(&self) -> bool {
+        match self {
+            Expr::ScalarFunction(ScalarFunction { func_def, .. }) => {
+                matches!(func_def, ScalarFunctionDefinition::BuiltIn(fun) if *fun == BuiltinScalarFunction::Coalesce)
+            }
+            Expr::BinaryExpr(BinaryExpr { op, .. }) => {
+                matches!(op, Operator::And | Operator::Or)
+            }
+            Expr::Case { .. } => true,
+            // Use explicit pattern match instead of a default
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:1281]: 👍

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @haohuaijin

**Comment by alamb** [comment]
Thank you @haohuaijin  and @jackwener  -- let's merge this PR and we can work on https://github.com/apache/arrow-datafusion/issues/8910 as a follow on

**Pull request merged**